feat: add withdraw logic to Collateral

ramenforbreakfast opened the pull request.

XavierEkkel reviewed (commented):
please adjust this PR with similar comments from https://github.com/prepo-io/prepo-monorepo/pull/496, and then re-request review!

ramenforbreakfast commented:
I have updated this PR with the following changes:
- Add ReentrancyGuard
- Added tests for different base tokens
- Changed revert case to follow the same format as `deposit` in the prior PR
- Changed event to follow the `Deposit` format

ramenforbreakfast commented:
Updated this PR once again, introducing changes that remove transfers of fee to treasury and instead provide approval to the `withdrawHook`. Also applied various minor nits that were mentioned in PR 503 as well.

XavierEkkel commented on apps/smart-contracts/core/contracts/Collateral.sol:
```diff
@@ -71,7 +71,27 @@ contract Collateral is
 
-  function withdraw(uint256 _amount) external override {}
+  function withdraw(uint256 _amount) external override nonReentrant {
+    uint256 _baseTokenAmount = (_amount * _baseTokenDenominator) / 1e18;
```
add comment above this line:
`Converts amount from collateral token units to base token units.`

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -834,4 +848,275 @@ describe('=> Collateral', () => {
+      await collateral.connect(deployer).setWithdrawHook(withdrawHook.address)
+    })
+
+    it('reverts if withdrawal 0 and fee rate 0', async () => {
```
same comments about tests here as in the deposit logic PR

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -834,4 +848,275 @@ describe('=> Collateral', () => {
+    })
+
+    it('reverts if hook reverts', async () => {
+      // Still providing valid inputs to ensure withdrawals are only reverting due to smock
```
cant you also check that the smock is called?

XavierEkkel commented on apps/smart-contracts/core/test/Collateral.test.ts:
```diff
@@ -834,4 +848,275 @@ describe('=> Collateral', () => {
+
+      await collateral.connect(user1).withdraw(amountToWithdraw)
+
+      expect(await baseToken.allowance(collateral.address, withdrawHook.address)).to.be.eq(fee)
```
add assertion that fee balance stays in the collateral contract

XavierEkkel reviewed (approved):
LGTM after comments
